### ISSUE TITLE
Persisting in-flight messages to disk

### ISSUE THREAD

**Issue by michael-p** [tekjar/rumq#47]
Dear all,

we're currently using the `rumqtt` crate but have hit kind of a road-block with that, and since that crate is archived we are considering switching to this crate.

The problem we're facing is that we need guaranteed message delivery, even when the software crashes or the system reboots, hence we must persist outgoing messages to disk and remove them only when we receive an acknowledgement from the broker.
There was some discussion in the (locked) issue tracker of `rumqtt` where the main author @tekjar  stated that they prefer to implement such a functionality outside of the crate. However, we were not able to do this, since there seems to be no way of knowing when a message is ack'ed by the broker.

So the question is: would this be possible to implement using `rumq`. Or, even better, does anybody have a tip on how to make this work using the "old" `rumqtt` crate? Sadly, we could not ask over there since the issue tracker is locked...

Best wishes, and thank you very much!
Michael

**Comment by tekjar** [comment]
> The problem we're facing is that we need guaranteed message delivery, even when the software crashes or the system reboots

When you say guaranteed message delivery, do you mean you cannot afford to loose data during crashes? That's going to be hard or inefficient even with persistence

**Comment by michael-p** [comment]
Well, it's not a problem if there is a window of a few seconds before the crash where we loose data but in principle users can turn of the device at any time, even when it's offline and in that case in-flight messages should be resend once the device starts up again and connectivity is restored.

Right now, using `rumqtt`, messages are queued in memory when the device is offline and then resend once it's back online again, but if the device reboots (or just the software restarts) during this offline-phase then messages are lost.

**Comment by tekjar** [comment]
We solve similar issues by plugging in  a disk aware stream before rumq

> users can turn of the device at any time

If you mean system shutdown by turn off, we flush everything to disk before shutdown

> it's offline and in that case in-flight messages should be resend once the device starts up again and connectivity is restored

If you plug a channel to feed requests into rumq's eventloop, you'll naturally get back pressure which indicates bad network. You can use it to orchestrate data between disk and memory

But if system randomly shuts down, you'll loose data. There is no way to prevent this unless you want to flush every message to disk before sending it network. Which is quite inefficient

**Comment by michael-p** [comment]
Thanks for your suggestions!

The way we tried to implement it so far (unsuccessfully) with `rumqtt` is very similar to what you describe: we persist every message to disk before handing it over to the rumqtt client.
The problem really was that there seems to be no way of knowing when the client successfully delivered a message to the broker so we can delete the message from disk. Could you elaborate on how back-pressure can be used here? We tried this (again, with `rumqtt`) but since the client has an internal in-flight queue with size at least 1 back-pressure is always at least 1 message "too late" (that is, even if there is no connection one can still send 1 message without getting back-pressure and hence one must assume the message was successfully delivered, which might not be the case).
Is this situation different with the new `rumq` implementation?

Our device sends messages only infrequently, so flushing every message to disk before sending is not a big deal. It looks like this is also the default behavior of Paho's disk-based persistence.

**Comment by tekjar** [comment]
Yeah unfortunately I'm not sure if there are zero sized channels ( used just for synchronization)

But probably write  1 message with QoS1 and wait for its ack? Only after receiving the previous message's ack, write the next message. Also remove the successfully acked message from disk?

**Comment by michael-p** [comment]
> But probably write 1 message with QoS1 and wait for its ack? Only after receiving the previous message's ack, write the next message. Also remove the successfully acked message from disk?

That would work, however at least with `rumqtt` we were unable to figure out how to wait for a message's ack (well, without directly modifying the eventloop:).
Is there a way to do this with `rumqtt` or `rumq`?

**Comment by tekjar** [comment]
This should be simple in rumq. Something like this

```rust
let stream = eventloop.stream();
requests.send(requests).await;

loop {
   received = request.next().await;
   if received == puback {
       break
   }
}
```

**Comment by michael-p** [comment]
Thanks again for the help! We have experimented a bit more with `rumq` and tried your idea and the new crate indeed looks very promising, good work!
However, we decided to stick with the "old" `rumqtt` for now as we're a small company and `rumqtt` still seems to be more stable and production-ready. But we will keep an eye on your work with `rumq` and might reconsider this in a few month.

Thanks again, and keep up the good work!
